src/odemis/driver/orsay.py

```diff
@@ -1389,3 +1440,1247 @@ def terminate(self):
+                             % parameter.Name)
+        if attributeName != "Actual":
+            return
+        state = self._ionColumn.ImageSize.Actual
+        logging.debug("Image format is: %s. Updating translation and resolution and their ranges accordingly." % state)
+        new_value = tuple(map(int, state.split(" ")))
+        self.imageFormat._value = new_value  # to not call the setter
+        self.imageFormat.notify(new_value)
+
+    def _translation_setter(self, value):
+        """
+        Setter of the translation VA.
+
+        :param (tuple (float, float)) value: Target translation of the area to image
+        :return (tuple (float, float)): The actual translation set
+
+        The translation VA marks the centre of the image area with respect to the centre of the field of view. This
+        setter transforms the coordinates of the centre of the image area to the coordinates of the top left corner of
+        the image area, which is the format the Orsay server takes. The setter also adjusts the size of the image area
+        (resolution VA) to prevent the new translation from placing part of the image area outside of the image format.
+        """
+        self.updatingImageArea.acquire()  # translation and resolution cannot be updated simultaniously
+        new_translation = list(value)
+
+        new_translation[0] = math.ceil(new_translation[0])
+        new_translation[1] = math.floor(new_translation[1])
+        if self.resolution.value[0] % 2 != 0:  # if horizontal resolution is odd
+            new_translation[0] -= 0.5  # prefer adding a pixel to the left
+        if self.resolution.value[1] % 2 != 0:  # if vertical resolution is odd
+            new_translation[1] += 0.5  # prefer adding a pixel to the top
+
+        # find the current limits for translation and clip the new value
+        tran_limit_0 = float(self.imageFormat.value[0] / 2 - self.resolution.value[0] / 2)
+        tran_limit_1 = float(self.imageFormat.value[1] / 2 - self.resolution.value[1] / 2)
+        if new_translation[0] < -tran_limit_0:
+            new_translation[0] = -tran_limit_0
+        elif new_translation[0] > tran_limit_0:
+            new_translation[0] = tran_limit_0
+        if new_translation[1] < -tran_limit_1:
+            new_translation[1] = -tran_limit_1
+        elif new_translation[1] > tran_limit_1:
+            new_translation[1] = tran_limit_1
+
+        target_translation = [0, 0]  # keep centre where it was, move target_trans from centre to upper left corner
+        target_translation[0] = int(self.imageFormat.value[0] / 2 + new_translation[0] - self.resolution.value[0] / 2)
+        target_translation[1] = int(self.imageFormat.value[1] / 2 - new_translation[1] - self.resolution.value[1] / 2)
+
+        target = map(str, target_translation + list(self.resolution.value))
+        target = " ".join(target)
+        self._ionColumn.ImageArea.Target = target
```

Comment:
All the code which is identical to the change of resolution should be shared.
Maybe have a function which get the new translation and resolution requested, and return the accepted translation and resolution, as they are set.